### PR TITLE
Changes to enable building msbuild with Unified Build

### PR DESCRIPTION
**Pull request by mmitche**
This enables building msbuild from within the VMR in non source-only modes.
- Disable building type libraries when building on Core, and looking to package them up.
- Bootstrap build should look for DotNetRoot, not RepoRoot. This will correctly pick up the global SDK from the VMR.
- Exclude a couple projects that are VS specific and cause issues (typically packaging TLBs)
- Don't look to copy net3.5 targeted PDBs when building on Core.

### CODE REVIEW

File: eng/Versions.props [hunk 1]
```diff
@@ -25,9 +25,29 @@
     <MicrosoftVSSDKBuildToolsVersion>16.7.13</MicrosoftVSSDKBuildToolsVersion>
   </PropertyGroup>
   <!-- Production Dependencies -->
+  <!-- There is some delicacy and balance between several scenarios here:
```
Comment (mmitche) [inline eng/Versions.props:28]: @rainersigwald let me know what you think of this solution.

**Comment by mmitche** [comment]
Confirmed on Linux and Windows,

File: src/Deprecated/Engine/Microsoft.Build.Engine.csproj [hunk 2]
```diff
@@ -4,7 +4,9 @@
     <NoWarn>$(NoWarn);618</NoWarn>
     <GenerateReferenceAssemblySource>true</GenerateReferenceAssemblySource>
     <EnableDefaultItems>false</EnableDefaultItems>
-    <CreateTlb>true</CreateTlb>
+    <!-- Do not create Tlbs when building in .NET product mode. The packages are not shipped to VS,
+         only their contents redisted within the SDK. -->
+    <CreateTlb Condition="'$(DotNetBuild)' != 'true'">true</CreateTlb>
```
Comment (JanKrivanek) [inline src/Deprecated/Engine/Microsoft.Build.Engine.csproj:9]: Where is the `DotNetBuild` defined?

File: src/Directory.Build.targets [hunk 3]
```diff
@@ -118,7 +118,10 @@
     </ItemGroup>
   </Target>
 
-  <Target Name="CreateTypeLib" BeforeTargets="AfterBuild" Inputs="$(TargetPath)" Outputs="$(TargetDir)$(TargetName).tlb;$(TargetDir)x64\$(TargetName).tlb" Condition="'$(BuildingInsideVisualStudio)' != 'true' and '$(CreateTlb)' == 'true' and $([MSBuild]::IsOSPlatform('windows')) and '$(TargetFrameworkIdentifier)' == '.NETFramework' and '$(MSBuildRuntimeType)' != 'Core'">
+  <!-- This target creates a type library for the assemblies in a package. It requires desktop msbuild to locate the tooling
+       Type libraries are still needed by Visual Studio in some cases, like the legacy C# project system.
+       It calls into msbuild via COM, and needs the type library to do so. -->
+  <Target Name="CreateTypeLib" BeforeTargets="AfterBuild" Inputs="$(TargetPath)" Outputs="$(TargetDir)$(TargetName).tlb;$(TargetDir)x64\$(TargetName).tlb" Condition="'$(BuildingInsideVisualStudio)' != 'true' and '$(CreateTlb)' == 'true' and $([MSBuild]::IsOSPlatform('windows')) and '$(TargetFrameworkIdentifier)' == '.NETFramework'">
```
Comment (JanKrivanek) [inline src/Directory.Build.targets:124]: Should the `'$(MSBuildRuntimeType)' != 'Core'` condition stay, to guard a case where `DotNetBuild` is not defined despite building on core?

**Review by JanKrivanek — APPROVED** [review]
Overall looks good - I'm just confused with conditions for type libs creation

File: eng/Packages.props [hunk 4]
```diff
@@ -23,13 +23,26 @@
...
     <PackageVersion Include="System.Text.Encoding.CodePages" Version="$(SystemTextEncodingCodePagesVersion)" />
     <PackageVersion Include="System.Text.Json" Version="$(SystemTextJsonVersion)" />
     <PackageVersion Include="System.Threading.Channels" Version="$(SystemThreadingChannelsVersion)" />
     <PackageVersion Include="System.Threading.Tasks.Dataflow" Version="$(SystemThreadingTasksDataflowVersion)" />
     <PackageVersion Include="xunit.console" Version="$(XUnitVersion)" />
+
+  </ItemGroup>
+
+  <!-- This set of packages will use different versions based on building binaries for VS. In non-UB modes this has no effect. -->
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework'">
```
Comment (ViktorHofer) [inline eng/Packages.props:38]: The logical expression `TFI == .NETFramework => Building binaries for VS` isn't true. It's very important to keep that in mind. I just last week had the same discussion on this with tmat and we landed on a nice design in Arcade: https://github.com/dotnet/arcade/pull/14391#discussion_r1465461640

File: eng/SourceBuild.props [hunk 5]
```diff
@@ -7,12 +7,13 @@
...
+  <Target Name="ConfigureInnerBuildArgs" BeforeTargets="GetSourceBuildCommandConfiguration">
     <PropertyGroup>
-      <InnerBuildArgs>$(InnerBuildArgs) /p:Projects="$(InnerSourceBuildRepoRoot)MSBuild.SourceBuild.slnf"</InnerBuildArgs>
-      <!-- Disable package validation as source build filters out target frameworks. -->
+      <!-- Filter down projects aggressively in source-only modes. -->
+      <InnerBuildArgs Condition="'$(ArcadeBuildFromSource)' == 'true' or '$(DotNetBuildSourceOnly)' == 'true'">$(InnerBuildArgs) /p:Projects="$(InnerSourceBuildRepoRoot)MSBuild.SourceBuild.slnf"</InnerBuildArgs>
+
+      <!-- Disable package validation as unified build may upgrade the net core target framework or
+           skip building net framework targets in source-only modes. -->
       <InnerBuildArgs>$(InnerBuildArgs) /p:EnablePackageValidation=false</InnerBuildArgs>
```
Comment (ViktorHofer) [inline eng/SourceBuild.props:17]: PackageValidation needs to be disabled when only producing a subset of TFMs. That only happens in source-only builds. I'm now missing a condition on this property. Why did you remove the target condition?

File: eng/Packages.props (hunk 4 above)
Comment (mmitche) [reply eng/Packages.props:38]: VS/packages really. IMO the key bit here is that is really the version flow problem I mentioned yesterday. We should not be transparently upgrading these dependencies except in source-only modes.

File: src/Deprecated/Engine/Microsoft.Build.Engine.csproj (hunk 2 above)
Comment (mmitche) [reply src/Deprecated/Engine/Microsoft.Build.Engine.csproj:9]: It comes in via Arcade, when any of the flags marking a VMR build is set.

File: eng/SourceBuild.props (hunk 5 above)
Comment (mmitche) [reply eng/SourceBuild.props:17]: I think this is due to the transparent upgrading of the target TFMs and the input packages.

Comment (ViktorHofer) [reply eng/SourceBuild.props:17]: We should disable the transparent upgrade of TFMs then and pin it in the repo. I assume msbuild just uses `NetCurrent`? This really needs to be fixed before the PR gets merged.

**Comment by mmitche** [comment]
@JanKrivanek @ViktorHofer I've changed the logic to not override the TFMs in normal UB modes.

File: eng/SourceBuild.props (hunk 5 above)
Comment (mmitche) [reply eng/SourceBuild.props]: Pinned in repo now.

File: eng/Packages.props (hunk 4 above)
Comment (rainersigwald) [reply eng/Packages.props]: @ViktorHofer I don't think I understand the concern. Can you elaborate?

File: src/Directory.Build.targets [hunk 6]
```diff
@@ -118,6 +118,9 @@
     </ItemGroup>
   </Target>
 
+  <!-- This target creates a type library for the assemblies in a package. It requires desktop msbuild to locate the tooling
+       Type libraries are still needed by Visual Studio in some cases, like the legacy C# project system.
+       It calls into msbuild via COM, and needs the type library to do so. -->
   <Target Name="CreateTypeLib" BeforeTargets="AfterBuild" Inputs="$(TargetPath)" Outputs="$(TargetDir)$(TargetName).tlb;$(TargetDir)x64\$(TargetName).tlb" Condition="'$(BuildingInsideVisualStudio)' != 'true' and '$(CreateTlb)' == 'true' and $([MSBuild]::IsOSPlatform('windows')) and '$(TargetFrameworkIdentifier)' == '.NETFramework' and '$(MSBuildRuntimeType)' != 'Core'">
```
Comment (rainersigwald) [inline src/Directory.Build.targets:124]: I saw you disabled this explicitly--was the MSBuildRuntimeType check here insufficient?

**Review by rainersigwald — APPROVED** [review]

File: src/Directory.Build.targets (hunk 6 above)
Comment (mmitche) [reply src/Directory.Build.targets:124]: You mean disabling via CreateTlb? There is use elsewhere in src/MSBuild/MSBuild.csproj https://github.com/dotnet/msbuild/pull/9685/files#diff-24f179152546e553997c333eeb9c000c2e31d5b3ff953b524fc49cdfc781fb6dR242

I thought it was clearer to have an explicit disabling of the creation rather than an implicit one based on the build host.

File: eng/Packages.props (hunk 4 above)
Comment (mmitche) [reply eng/Packages.props]: The changes here have since been reverted. in non source-only modes, msbuild builds very very close to how it builds in the isolated repo context.

File: eng/SourceBuild.props [hunk 7]
```diff
@@ -7,11 +7,11 @@
     <SourceBuildManagedOnly>true</SourceBuildManagedOnly>
   </PropertyGroup>
 
-  <Target Name="ConfigureInnerBuildArgs"
-          BeforeTargets="GetSourceBuildCommandConfiguration"
-          Condition="'$(ArcadeBuildFromSource)' == 'true'">
-    <PropertyGroup>
+  <Target Name="ConfigureInnerBuildArgs" BeforeTargets="GetSourceBuildCommandConfiguration">
+    <PropertyGroup Condition="'$(ArcadeBuildFromSource)' == 'true' or '$(DotNetBuildSourceOnly)' == 'true'">
```
Comment (ViktorHofer) [inline eng/SourceBuild.props:11]: Move the condition back into the target and revert putting everything into a single line. We find separate lines more readable for msbuild targets that have more than two xml attributes.

File: eng/Packages.props (hunk 4 above)
Comment (ViktorHofer) [reply eng/Packages.props]: > @ViktorHofer I don't think I understand the concern. Can you elaborate?

Pinning package versions globally to a specific version for VS support based on the `$(TargetFramework` or `$(TargetFrameworkIdentifier)` is an incorrect assumption. You could have a normal class library that uses i.e. System.Text.Json and multi-targets for both .NET Framework and .NETCoreApp. Pinning the System.Text.Json version down to X for the .NET Framework build doesn't make sense. VS support applies to msbuild tasks, etc but not to class libraries that ship out-of-band via nuget packages.

My general recommendation around this has been to mark projects that apply to these restrictions via some property, i.e. https://github.com/dotnet/arcade/blob/2d7ad177f759e8a0128cc53fa90ffc54aec1822c/src/Microsoft.DotNet.NuGetRepack/tasks/Microsoft.DotNet.NuGetRepack.Tasks.csproj#L6.

In the example of Arcade, it then imports a targets file that updates the PackageVersion items for VS support: https://github.com/dotnet/arcade/blob/2d7ad177f759e8a0128cc53fa90ffc54aec1822c/eng/BuildTask.targets#L3

**Review by ViktorHofer — APPROVED** [review]

File: eng/Packages.props (hunk 4 above)
Comment (rainersigwald) [reply eng/Packages.props]: > VS support applies to msbuild tasks, etc but not to class libraries that ship independently.

I think I agree with this but I don't see how it applies to this repo, which is filled with things where we only care about the .NET Framework versions of things because of VS specifically.
